integrations/uniswap_v4.py

```diff
@@ -0,0 +1,246 @@
+    ownerOf_calls = []
+    token_ids = []
+
+    while start < end:
+        to_block = start + PAGINATION_SIZE
+        deposits = fetch_events_logs_with_retry(
+            "Users",
+            uniswap_v4_pm_contract.events.ModifyLiquidity(),
+            start,
+            to_block,
+        )
+        deposits_list = list(deposits)
+        print(start, to_block, len(deposits_list), "getting deposits for uniswap v4")
+
+        for deposit in deposits_list:
+            if (
+                deposit["args"]["id"] == UNISWAP_V4_USDE_POOL
+                and deposit["args"]["sender"] == UNISWAP_V4_NFPM_ADDRESS
+            ):
+                token_id = Web3.to_int(deposit["args"]["salt"])
+
+                ownerOf_calls.append(
+                    (
+                        uniswap_v4_nfpm_contract,
+                        uniswap_v4_nfpm_contract.functions.ownerOf.fn_name,
+                        [token_id],
+                    )
+                )
+                token_ids.append(token_id)
+
+        print(start, to_block, len(token_ids), "found valid token ids")
+        start += PAGINATION_SIZE
+
+    # Execute multicall for all token ownerships
+    if ownerOf_calls:
+        multicall_results = multicall_by_address(
+            wb3=w3,
+            multical_address="0xcA11bde05977b3631167028862bE2a173976CA11",
+            calls=ownerOf_calls,
+            block_identifier=end,
+            allow_failure=True,
+        )
+
+        # Process results
+        for token_id, result in zip(token_ids, multicall_results):
+            if result == None:
+                continue
+
+            try:
+                owner = Web3.to_checksum_address(result[0])
+            except Exception:
+                continue
+
+            if owner not in users:
+                users[owner] = {token_id}
+            else:
+                users[owner].add(token_id)
+
+    return users
+
+
+def get_position_balance(token_id, block, tick, sqrt_price):
+    try:
+        packedInfo = call_with_retry(
+            uniswap_v4_nfpm_contract.functions.positionInfo(token_id),
+            block,
+        )
+        liquidity = call_with_retry(
+            uniswap_v4_nfpm_contract.functions.getPositionLiquidity(token_id),
+            block,
+        )
+    except Exception:
+        print(f"token {token_id} empty at block {block}")
+        return [0, 0]
+
+    if liquidity == 0:
+        return [0, 0]
+
+    t0 = 0
+    t1 = 0
+
+    # tickUpper: bits 32-55 (from the right, 0-indexed)
+    # tickLower: bits 8-31
+    tick_upper = extract_signed_24bit((packedInfo >> 32) & ((1 << 24) - 1))
+    tick_lower = extract_signed_24bit((packedInfo >> 8) & ((1 << 24) - 1))
+
+    # print(
+    #     f"tick: {tick}, tick_lower: {tick_lower}, tick_upper: {tick_upper}, token_id: {token_id}, liquidity: {liquidity}"
+    # )
+
+    sqrt_ratio_l = math.sqrt(1.0001**tick_lower)
+    sqrt_ratio_u = math.sqrt(1.0001**tick_upper)
```

Comment:
seems like these need to be scaled in Q96 to be safely compared to `sqrt_price`

Reply:
Not enough, that actually doesn't change the output by much surprisingly.